src/webview/java/org/medicmobile/webapp/mobile/EmbeddedBrowserActivity.java

```diff
@@ -435,6 +451,26 @@ public void onReceivedError(WebView view, WebResourceRequest request, WebResourc
 							"}", error.getErrorCode(), error.getDescription()), false);
 				}
 			}
+
+			@Override
+			public void onPageStarted(WebView view, String url, Bitmap favicon) {
+				trace(this, "onPageStarted() :: url: %s, isMigrationRunning: %s", url, isMigrationRunning);
+				if (isMigrationRunning && url.contains("/login")) {
+					isMigrationRunning = false;
+					CookieManager cookieManager = CookieManager.getInstance();
+					String cookie = cookieManager.getCookie(appUrl);
+					if (cookie == null) {
+						log(this, "onPageStarted() :: Migration process in progress, and " +
+								"cookies were not loaded, restarting ...");
+						Context context = view.getContext();
+						Intent intent = new Intent(context, StartupActivity.class);
+						intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
+						context.startActivity(intent);
+						Runtime.getRuntime().exit(0);
```

Comment:
This section is really important, and quite difficult to understand. Consider pulling out a private function that can be named clearly. In addition it probably needs some commenting about what this is doing and why it works.